toothpick-kotlin/src/main/java/toothpick/kotlin/ToothpickInjectDelegate.kt

```diff
@@ -0,0 +1,45 @@
+package toothpick.kotlin
+
+import toothpick.Scope
+import kotlin.properties.ReadOnlyProperty
+import kotlin.reflect.KClass
+import kotlin.reflect.KProperty
+
+typealias InputConfigBlock = (InjectConfig.() -> Unit)
+
+open class ToothpickInjectDelegate<OWNER, T : Any>(
+    protected val clazz: KClass<T>,
+    protected val scope: Scope?,
+    protected val name: String? = null,
+    protected val annotationName: KClass<out Annotation>? = null,
+    protected val injectConfigBlock: InputConfigBlock? = null
+) : ReadOnlyProperty<OWNER, T> {
+    private var value: T? = null
+
+    override fun getValue(thisRef: OWNER, property: KProperty<*>): T {
+        return value ?: injectValue(thisRef, property)
+    }
+
+    protected open fun injectValue(thisRef: OWNER, property: KProperty<*>): T {
+        val injectConfig = InjectConfig()
+        if (injectConfigBlock != null) {
+            injectConfig.apply(injectConfigBlock)
+        }
+        val scope = injectConfig.scope ?: scope ?: getScope(thisRef, property) ?: throw IllegalStateException("No valid scope available when attempting to inject $property in $thisRef")
+        return scope.getInstance(clazz, injectConfig.name ?: name, injectConfig.annotationName ?: annotationName).also { value = it }
+    }
+
+    protected open fun getScope(thisRef: OWNER, property: KProperty<*>): Scope? {
+        if (scope != null) {
+            return scope
+        }
+
+        if (thisRef is HasScope) {
+            return thisRef.scope
+        }
```

Comment:
I got an issue with this. No idea if it is a standard way to implement other DI engines usually. 
But it seems very hacky for a few reasons. I understand that we have this mechanism, and the fallback of creating a new scope right under the app for activities, because we want to be more concise and not pass the scope when doing `by inject(scope)`

But first things first, I am not sure I understand how lazy and inject differ now. Isn't the delegated property going to be "computed" only when the code accesses it ? if so, what's the difference with a lazy injection, and if not, when is it gonna be computed?

I don't really like the interface, and the fallback neither. It assumes that the activity as a given property, implements a given interface. An activity could decide to call it's scope `scopeA` and forget about the interface and no injection would take place.
Same for the fallback, we don't want to consider the activity's scope is placed right under the application by default. And for fragments, I am not even sure that the activity scope should be the default.

I was thinking that it would be maybe cleaner:
1) to pass the scope all the time when injecting / lazying a property. `by inject(scope)`. We could also prefer to use a scope name `by inject(this)`
2) the activity can itself have a getter/delegate to create it's scope when needed
```kotlin
  lateinit var scope: Scope
     get() { 
            TP.createScope(...)
            scope.installModules(...)
            return scope
    }
```
3) if it's not convenient to pass the scope to every inject call. Can't we have another mechanism where we would restor the call to TP.inject(this, scope) and this would provide the scope to `by inject()` 's delegate?

Reply:
And there is another issue. IMHO. We're loosing the `@Inject` annotation for fields. It seems to me there is a tension in your PR between what happened to butterknife (replaced by lateinit `by findViewByID`) and the classic `@Inject` annnotation. I prefer the annotation, but why did you prefer the  other way, for speed ?

Reply:
I'm not a fan of this hack. I haven't found a way to do this better without passing scope with every call. At one point, I had scope as nullable, so if you didn't provide it the delegate would try to use `this`. I didn't really like that either but it may be better than this route.

In most cases, I don't want to have to pass scope ever unless I'm doing something fancy.

I like no longer having the `@Inject` annotations, less non-code in my code to read around. We gain a few things with it:
1. No longer need an annotation processor, quicker build time
2. No longer need member injectors, quicker build time, no java leaking into pure Kotlin modules causing more compilation
3. Private injected members, no need to weaken access modifier to allow an APT generated class access. 
4. No longer need to set as `lateinit var` or nullable
5. IMO, reads cleaner
6. Puts all injection logic in one spot as you can put scope and name in there as well, it's all done in one go instead of multiple annotations or split of logic.

Reply:
1. we still need the factories to be created.
2. factories could be implemented in kotlin using extension function but I don't know any AP doing this. Not sure the build times would be any better.
3. that one ok
4. I am really motivated by keystroke count.`@Inject lateinit var foo` vs `lateinit var foo by inject()`
5...
6. It also changes the logic of the instantiation and scopes. Now the scope would be decided by the client whereas in java it's set by the entity itself. I feel that it might be conflicting in some cases too.

Reply:
1. Yup, fair point for some classes, others won't need it like activities and fragments (nice for Android land, not applicable for Java land).
2. Sure, factories are still needed for most, and there is still Java leakage, but less and less to get rid of if TP ever moves to a kotlin based solution.
3. A big plus in my eyes as it influences 4
4. Actual code:
`@Inject lateinit var foo: Foo`
`private val foo: Foo by inject()`

I really like the second vs the first, and in the day of IDEs and code complete, keystroke count alone is a poor argument loosening access restrictions and adding mutability to the variable. Big philosophy of Kotlin is immutability, and this change moves TP in that direction.

5. -
6. In my use cases, it hasn't been much of a big deal as I rarely am doing anything special with scopes. And when I am, I'm already paying extra attention to how I'm using the scope. I'm not sure I follow what you mean by `client` and `entity`, could you elaborate?

Reply:
2. Do kotlin people really talk about java leakage ? I find it funny ;)

I get it for #4. That's a good point (sorry I am really new to kotlin, it takes some time to put the pieces together ;)). Yes for immutability, it's a big win. But I like my so much my little cute annotations ! Snifff...

6. I understand why the PR goes a little too far with scopes now.  ;)